File: MultiThread/Semaphore.py
from threading import Thread
class Semaphore:
        def __init__(self,s):
                self.s = s
        def wait(self):
                while(self.s==False):
                        '''Waits'''
                self.s = False
        def signal(self):
                self.s = True
if __name__ == '__main__':
        s0 = Semaphore(True)
        s1 = Semaphore(False)
        s2 = Semaphore(False)
        def th1():
                s1.wait()
                s0.signal()
        def th2():
                s2.wait()
                s0.signal()
        def th0():
                while(True):
                        s0.wait()
                        print 0
                        s1.signal()
                        s2.signal()
        t0 = Thread(target=th0)
        t1 = Thread(target=th1)
        t2 = Thread(target=th2)
        t0.start()
        t1.start()
        t2.start()

File: MultiThread/Semaphore.cpp
#include <iostream>
#include <thread>  
//Compile : g++ -pthread 
using namespace std;
class Semaphore {
	private:
		bool s;
	public:
		Semaphore(bool s) {
			this->s = s;
		}
		void Wait() {
			while(!s);
			s = false;
		}
		void Signal() {
			s = true;
		}
};
Semaphore s0(true) ,s1(false), s2(false);
void f0() {
	while(!false) {
		s0.Wait();
		cout<<"0";
		s1.Signal();
		s2.Signal();
	}
}
void f1() {
	s1.Wait();	
	s0.Signal();
}
void f2() {
	s2.Wait();	
	s0.Signal();
}
int main() {
	thread t1(f1),t2(f2),t0(f0);
	t1.join();
	t2.join();
	t0.detach();
	return 0;
}
